codespell: config, workflow and some typo fixes

yarikoptic opened the pull request:
Excluded .py files entirely for this initial pass. Someone can remove skipping them and improve there as well

yarikoptic committed:
Add github action to codespell develop on push and PRs

yarikoptic committed:
Add rudimentary .codespellrc

yarikoptic committed:
For now exclude .py recipes

yarikoptic committed:
ignores

yarikoptic committed:
Fix ambigous typos

yarikoptic committed:
more ignores

wdconinc commented:
Rebased and conflicts resolved (file removed).

wdconinc commented on .github/workflows/codespell.yml:
```diff
@@ -0,0 +1,19 @@
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Codespell
+        uses: codespell-project/actions-codespell@v1
```
Use commit hash, upgrade to v2.
```suggestion
        uses: codespell-project/actions-codespell@94259cd8be02ad2903ba34a22d9c13de21a74461
```

wdconinc commented on .github/workflows/codespell.yml:
```diff
@@ -0,0 +1,19 @@
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
```
Use commit hash.
```suggestion
        uses: actions/checkout@44c2b7a8a4ea60a981eaca3cf939b5f4305c123b
```

wdconinc reviewed (commented):
Thanks for this PR! Apologies for the delayed review.

Your PR is reasonable scoped (i.e. doesn't include packages at this stage), which I think is the right approach.

wdconinc commented on var/spack/repos/builtin/packages/libsigsegv/patch.new_config_guess:
```diff
@@ -113,7 +113,7 @@
      *:NetBSD:*:*)
  	# NetBSD (nbsd) targets should (where applicable) match one or
--	# more of the tupples: *-*-netbsdelf*, *-*-netbsdaout*,
+-	# more of the tuples: *-*-netbsdelf*, *-*-netbsdaout*,
```
Can you please remove this file from the changeset? If we fix the spelling in patches, then we will not be able to apply them.

yarikoptic committed:
Review: use specific commit hashes instead of version branches for actions

Co-authored-by: Wouter Deconinck <wdconinc@gmail.com>

yarikoptic committed:
Do not codespell fix patch and spack.yaml files

yarikoptic committed:
[DATALAD RUNCMD] Do interactive fixing of some ambigous typos

=== Do not change lines below ===
{
 "chain": [],
 "cmd": "codespell -w -i 3 -C 2",
 "exit": 0,
 "extra_inputs": [],
 "inputs": [],
 "outputs": [],
 "pwd": "."
}
^^^ Do not change lines above ^^^

yarikoptic committed:
Ignore capitalized THIRDPARTY

yarikoptic committed:
[DATALAD RUNCMD] run codespell throughout fixing typos automagically

=== Do not change lines below ===
{
 "chain": [],
 "cmd": "codespell -w",
 "exit": 0,
 "extra_inputs": [],
 "inputs": [],
 "outputs": [],
 "pwd": "."
}
^^^ Do not change lines above ^^^

yarikoptic commented:
ok, should be good now -- ignored `patches.*` altogether, and did few other tune ups and rebased again so it is all uptodate.

wdconinc commented on .codespellrc:
```diff
@@ -0,0 +1,5 @@
+[codespell]
+skip = .git,*.pdf,*.svg,*.patch,*.py,*.diff,*.pub,foamEtcFile,patch.*
```
Is there a way to skip the intentional non-utf-8 test files like `lib/spack/spack/test/data/filter_file/x86_cpuid_info.c`? It flags them with an annotation at e.g. https://github.com/spack/spack/actions/runs/9159378954.

We probably also should ignore `.dat` files (one is flagged as well) but I've filed a feature request to remove that particular one.

yarikoptic replied:
I don't think there is a way to entirely skip the files which do not decode as utf-8 or to specify which encoding to use per file . ATM it is hardcoded to try        `for encoding in ("utf-8", "iso-8859-1"):` . And thus it does catch/fix typos already.  We can add pragma to ignore specific lines, but that wouldn't ignore entire file reading first.

Filed
- https://github.com/codespell-project/codespell/issues/3427

and feel welcome to comment on if you see some other desired way.

re .dat -- sure we could ignore those if you like. pushed

yarikoptic replied:
pushed commit to silence all of those!

yarikoptic committed:
Also ignore .dat to avoid decoding problems

yarikoptic committed:
Adjust quiet level to not spit out warnings for encoding issues

alecbcs commented:
I would push for the new workflow be an extra step inside of the existing style checks rather than its own standalone CI. 
(Or at least reuse our existing infrastructure inside of [workflows/ci.yml](https://github.com/spack/spack/blob/develop/.github/workflows/ci.yaml) as an additional precheck.)

wdconinc commented:
> I would push for the new workflow be an extra step inside of the existing style checks rather than its own standalone CI.

Suggestion: add to https://github.com/spack/spack/blob/develop/.github/workflows/valid-style.yml after `validate`, before `style`?

adamjstewart commented:
I kind of wonder if typo checking is more useful as a warning on release branches than as an error on every commit. I'm still concerned about false positives, but I guess we would have to run it on the entire repo to see how many of those there are.

yarikoptic commented:
FWIW, we often have codespell even in pre-commit configuration so no typos even get committed. IMHO the sooner they are defected and fixed the better. 
False positives can happen but rare and can be whitelisted. 

Decide where to plug the scrub and I will do it feel free to push such a commit
